File: Questions/prog_<3>.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data ;
    struct node *next ;
};

struct node *createnode(int new_data){
    struct node *new_node = (struct node*)malloc(sizeof(struct node));
    new_node -> data = new_data;
    new_node -> next = NULL;
    return new_node;
}

// Fixed return type from Node to node
struct node *reverse_traversal(struct node *head){
    if (head == NULL) return NULL; // Handle empty list

    struct node *stack [1000];
    int top = -1;
    struct node *temp = head ;
    
    while (temp != NULL){
        stack [++top] = temp ;
        temp = temp ->next;
    }

    if (top >= 0){
        head = stack [top];
        temp = head ;
        while (top > 0){
            temp -> next = stack[--top];
            temp = temp -> next ;
        }
        temp -> next = NULL ;
    }
   return head;
}

void print_list(struct node *node){
    while (node != NULL){
        printf("%d",node -> data);
        if (node->next){
            printf("->");
        }
        node = node->next;
    }
    printf("\n");
}

int main() {
    // Fixed type naming in main to match struct node
    struct node *head = createnode(1);
    head->next = createnode(2);
    head->next->next = createnode(3);
    head->next->next->next = createnode(4);
    head->next->next->next->next = createnode(5);

    head = reverse_traversal(head);
    print_list(head);
    return 0;
}

File: Questions/prog_<3>.py
class node:
    def __init__(self, data):
        self.data = data
        self.next = None

class linkedlist:
    def __init__(self, head):
        self.head = head

def reverse_traversal(head):
    current = head
    prev = None

    while current is not None:
        NextNode = current.next
        current.next = prev

        prev = current
        current = NextNode
    return prev


def main(node):
    while node is not None:
        print(node.data, end="")
        if node.next is not None:
            print("->", end="")
        node = node.next
    print()


if __name__ == "__main__":
    head = node(1)
    head.next = node(2)
    head.next.next = node(3)
    head.next.next.next = node(4)
    head.next.next.next.next = node(5)

    head = reverse_traversal(head)
    main(head)
